plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,312 @@
+                        .getAdditionalProperties()
+                        .put(field.getName(), String.valueOf(field.get(queryStatistics)));
+            }
+            return Optional.of(trinoQueryStatisticsFacet);
+        }
+        return Optional.empty();
+    }
+
+    private OpenLineage.RunEvent getStartEvent(UUID runID, QueryCreatedEvent queryCreatedEvent)
+            throws IllegalAccessException
+    {
+        OpenLineage.RunFacetsBuilder runFacetsBuilder = ol.newRunFacetsBuilder();
+        Optional<OpenLineage.RunFacet> trinoMetadata = getTrinoMetadataFacet(queryCreatedEvent.getMetadata());
+        Optional<OpenLineage.RunFacet> trinoQueryContext = getTrinoQueryContextFacet(queryCreatedEvent.getContext());
+
+        trinoMetadata.ifPresent(runFacet -> runFacetsBuilder.put("trino.metadata", runFacet));
+        trinoQueryContext.ifPresent(runFacet -> runFacetsBuilder.put("trino.queryContext", runFacet));
+
+        OpenLineage.RunEvent startEvent =
+                ol.newRunEventBuilder()
+                        .eventType(OpenLineage.RunEvent.EventType.START)
+                        .eventTime(queryCreatedEvent.getCreateTime().atZone(ZoneId.of("UTC")))
+                        .run(ol.newRunBuilder().runId(runID).facets(runFacetsBuilder.build()).build())
+                        .job(
+                                ol.newJobBuilder()
+                                        .namespace(getJobNamespace())
+                                        .name(queryCreatedEvent.getMetadata().getQueryId())
+                                        .facets(
+                                                ol.newJobFacetsBuilder()
+                                                        .sql(ol.newSQLJobFacet(queryCreatedEvent.getMetadata().getQuery()))
+                                                        .build())
+                                        .build())
+                        .build();
+
+        return startEvent;
+    }
+
+    private OpenLineage.RunEvent getCompletedEvent(UUID runID, QueryCompletedEvent queryCompletedEvent)
+            throws IllegalAccessException
+    {
+        boolean failed = queryCompletedEvent.getMetadata().getQueryState().equals("FAILED");
+
+        OpenLineage.RunFacetsBuilder runFacetsBuilder = ol.newRunFacetsBuilder();
+        Optional<OpenLineage.RunFacet> trinoMetadata = getTrinoMetadataFacet(queryCompletedEvent.getMetadata());
+        Optional<OpenLineage.RunFacet> trinoQueryStatistics = getTrinoQueryStatisticsFacet(queryCompletedEvent.getStatistics());
+        Optional<OpenLineage.RunFacet> trinoQueryContext = getTrinoQueryContextFacet(queryCompletedEvent.getContext());
+
+        trinoMetadata.ifPresent(runFacet -> runFacetsBuilder.put("trino.metadata", runFacet));
+        trinoQueryStatistics.ifPresent(runFacet -> runFacetsBuilder.put("trino.queryStatistics", runFacet));
+        trinoQueryContext.ifPresent(runFacet -> runFacetsBuilder.put("trino.queryContext", runFacet));
+
```

Comment:
Instead of emitting two lineage nodes, one for `START` and one for `COMPLETE`, can we not just rely on `QueryCompletedEvent` and send just one `COMPLETE` event if query execution was successful.

Also to expand my other follow up above around mutation queries, If It is a `CTAS` query that failed, do we really need to emit a `START` event and `FAIL` event or just skip sending any event because query has failed.

Reply:
in my opinion we shouldn't leave out START events as:
- state change START > COMPLETE can be a source of valuable information
- having both states/events leaves more audit information

@mobuchowski do you have an opinion about just leaving COMPLETE events? are there any OL integrations leaving START events on purpose?

Reply:
>are there any OL integrations where it makes sense to leave out START events on purpose?

Not on purpose. If it's possible to get `START` events - and I see it is - it would definitely be preferable for OpenLineage backends - @mgorsk1 is right.

>If It is a CTAS query that failed, do we really need to emit a START event and FAIL event or just skip sending any lineage event because query has failed.

OpenLineage backends can also take advantage of FAIL events, it's the same story - additional information.